src/messages.rs

```diff
@@ -0,0 +1,299 @@
+///
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+enum Payload {
+    /// KeyExchange take the following form:
+    ///
+    /// | Field      | Length | Description                                                        |
+    /// |------------|--------|--------------------------------------------------------------------|
+    /// | Signature  | 64     | Compressed ECDSA signature over the following public key using the |
+    /// |            |        | long term  public key                                              |
+    /// | Public Key | 33     | Compressed ephemeral public key                                    |
+    ///
+    KeyExchange {
+        signature: Signature,
+        public_key: PublicKey,
+    },
+}
+
+fn from_bytes_keyexchange<'a>(
+    buf: &'a [u8],
+    secp256k1: &Secp256k1,
+) -> io::Result<Option<(Payload, &'a [u8])>> {
+    if buf.len() < 64 + 33 {
+        return Ok(None);
+    }
+    let (signature, buf) = buf.split_at(64);
+    let signature = {
+        match Signature::from_compact(secp256k1, signature) {
+            Ok(sig) => sig,
+            Err(_) => {
+                return Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    "Signature parsing failed",
+                ))
+            }
+        }
+    };
+    let (public_key, buf) = buf.split_at(33);
+    let public_key = {
+        match PublicKey::from_slice(secp256k1, public_key) {
+            Ok(pk) => pk,
+            Err(_) => {
+                return Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    "Public key parsing failed",
+                ))
+            }
+        }
+    };
+    return Ok(Some((
+        Payload::KeyExchange {
+            signature: signature,
+            public_key: public_key,
+        },
+        buf,
+    )));
+}
+
+
+fn to_bytes_keyexchange(
+    signature: &Signature,
+    public_key: &PublicKey,
+    buf: &mut BytesMut,
+    secp256k1: &Secp256k1,
+) -> io::Result<()> {
+    buf.extend(vec![0x00]);
+    buf.extend(signature.serialize_compact(secp256k1).iter());
+    buf.extend(public_key.serialize_vec(secp256k1, true));
+    Ok(())
+}
+
+impl Payload {
+    /// Takes a byte slice and builds a Payload. Returns same as MessageCodec::decode plus the
+    /// remaining (unconsumed) buffer.
+    fn from_bytes<'a>(
+        buf: &'a [u8],
+        secp256k1: &Secp256k1,
+    ) -> io::Result<Option<(Payload, &'a [u8])>> {
+        if buf.len() < 1 {
+            return Ok(None);
+        }
+        let (variant, buf) = buf.split_at(1);
+        match variant[0] {
+            0 => return from_bytes_keyexchange(buf, secp256k1),
+            _ => {
+                return Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    "Unknown payload variant",
+                ))
+            }
+        }
+    }
+    fn to_bytes(&self, buf: &mut BytesMut, secp256k1: &Secp256k1) -> io::Result<()> {
```

Comment:
Also here... looks like we will have a lot of boilerplate matches with the enum solution.